### PR TITLE
feat: support black box test

### PR DESCRIPTION
**Pull request by Young-Flash**
support *_bbtest.mbt as black box test

### CODE REVIEW

File: crates/moon/tests/test_cases/mod.rs [hunk 1]
```diff
@@ -4128,3 +4128,63 @@ fn test_third_party() {
...
+            moonc build-package ./A/hello.mbt -o ./target/wasm-gc/debug/test/A/A.core -pkg username/hello/A -std-path $MOON_HOME/lib/core/target/wasm-gc/release/bundle -pkg-sources username/hello/A:./A -target wasm-gc -g
+            moonc build-package ./C/hello.mbt -o ./target/wasm-gc/debug/test/C/C.core -pkg username/hello/C -std-path $MOON_HOME/lib/core/target/wasm-gc/release/bundle -pkg-sources username/hello/C:./C -target wasm-gc -g
+            moonc build-package ./B/hello.mbt -o ./target/wasm-gc/debug/test/B/B.core -pkg username/hello/B -std-path $MOON_HOME/lib/core/target/wasm-gc/release/bundle -pkg-sources username/hello/B:./B -target wasm-gc -g
+            moonc build-package ./A/hello_bbtest.mbt ./target/wasm-gc/debug/test/A/__generated_driver_for_blackbox_test.mbt -o ./target/wasm-gc/debug/test/A/A.blackbox_test.core -pkg username/hello/A -is-main -std-path $MOON_HOME/lib/core/target/wasm-gc/release/bundle -i ./target/wasm-gc/debug/test/A/A.mi:A -i ./target/wasm-gc/debug/test/C/C.mi:C -i ./target/wasm-gc/debug/test/B/B.mi:B -pkg-sources username/hello/A:./A -target wasm-gc -g
```
Comment (lijunchen) [inline crates/moon/tests/test_cases/mod.rs]: `B` is a `test_import` in package `A`. The command `moonc build-package ./A/hello_bbtest.mbt ...` should not contain `-i ./target/wasm-gc/debug/test/B/B.mi:B`.

File: crates/moon/tests/test_cases/mod.rs [hunk 2]
```diff
@@ -4128,3 +4128,61 @@ fn test_third_party() {
...
+            &dir,
+            [
+                "test",
+                "-p",
+                "username/hello/A",
+                "-f",
+                "hello_bbtest.mbt",
+                "-i",
+                "0",
+                "--dry-run",
```
Comment (lijunchen) [inline crates/moon/tests/test_cases/mod.rs]: For `--dry-run`, we can use `--nostd` to simplify the output

File: crates/moon/tests/test_cases/blackbox_success_test.in/A/moon.pkg.json [hunk 3]
```diff
@@ -1,8 +1,5 @@
 {
   "import": [],
-  "test_import": [
-    "username/hello/B"
```
Comment (lijunchen) [inline crates/moon/tests/test_cases/blackbox_success_test.in/A/moon.pkg.json]: Why delete this?

File: crates/moon/tests/test_cases/mod.rs [hunk 4]
```diff
@@ -4151,21 +4151,19 @@ fn test_blackbox_success() {
...
+            moonc build-package ./A/hello_bbtest.mbt ./target/wasm-gc/debug/test/A/__generated_driver_for_blackbox_test.mbt -o ./target/wasm-gc/debug/test/A/A.blackbox_test.core -pkg username/hello/A -is-main -std-path $MOON_HOME/lib/core/target/wasm-gc/release/bundle -i ./target/wasm-gc/debug/test/A/A.mi:A -i ./target/wasm-gc/debug/test/C/C.mi:C -pkg-sources username/hello/A:./A -target wasm-gc -g
+            moonc link-core $MOON_HOME/lib/core/target/wasm-gc/release/bundle/core.core ./target/wasm-gc/debug/test/A/A.core ./target/wasm-gc/debug/test/C/C.core ./target/wasm-gc/debug/test/A/A.blackbox_test.core -main username/hello/A -o ./target/wasm-gc/debug/test/A/A.blackbox_test.wasm -test-mode -pkg-sources username/hello/C:./C -pkg-sources username/hello/A:./A -pkg-sources moonbitlang/core:$MOON_HOME/lib/core -target wasm-gc -g
```
Comment (lijunchen) [inline crates/moon/tests/test_cases/mod.rs]: If you test-import B, `moonc link-core` will link an unused `B.core` for `A.blackbox_test.core`

File: crates/moonbuild/src/gen/gen_runtest.rs [hunk 5]
```diff
@@ -240,8 +312,14 @@ fn get_pkg_topo_order<'a>(m: &'a ModuleDB, leaf: &Package) -> Vec<&'a Package> {
         }
         visited.insert(cur_pkg_full_name.clone());
         let cur_pkg = &m.packages[cur_pkg_full_name];
+        // TODO: don't always chain test_imports and bbtest_imports?
```
Comment (lijunchen) [inline crates/moonbuild/src/gen/gen_runtest.rs]: This needs to be fixed.

See comment at crates/moon/tests/test_cases/mod.rs:4155

File: crates/moonbuild/src/gen/gen_runtest.rs [hunk 6]
```diff
@@ -225,6 +225,78 @@ pub fn gen_package_underscore_test(
     })
 }
 
+pub fn gen_package_blackbox_test(
```
Comment (lijunchen) [inline crates/moonbuild/src/gen/gen_runtest.rs:228]: Do we really need this hard-coded rule similar to the internal test for the black box test?

I once thought that if we encapsulated `*_bbtest.mbt` files into a `moonutil::Package`, we could get black box tests for free. However, I find the hard part is generating the test driver.

File: crates/moon/tests/test_cases/mod.rs (hunk 4 above)
Comment (Young-Flash) [reply crates/moon/tests/test_cases/mod.rs]: yeah, the previous `cur_pkg.imports.iter().chain(cur_pkg.test_imports.iter())` will result in link an unused `test_import.core` for  `*.internal_test.core`

File: crates/moon/tests/test_cases/blackbox_success_test.in/A/moon.pkg.json (hunk 3 above)
Comment (Young-Flash) [reply crates/moon/tests/test_cases/blackbox_success_test.in/A/moon.pkg.json]: no use in the actual code.

but can preserve to make sure `test_import.core` shouldn't be link into `blackbox_test.core`

File: crates/moonbuild/src/gen/gen_runtest.rs (hunk 6 above)
Comment (lijunchen) [reply crates/moonbuild/src/gen/gen_runtest.rs:228]: But the current solution also LGTM

File: crates/moonbuild/src/gen/gen_runtest.rs [hunk 7]
```diff
@@ -296,50 +299,61 @@ pub fn gen_package_blackbox_test(
...
+    with_bbtest_import: bool,
+) -> Vec<&'a Package> {
     let mut visited: HashSet<String> = HashSet::new();
     let mut pkg_topo_order: Vec<&Package> = vec![];
     fn dfs<'a>(
         m: &'a ModuleDB,
         pkg_topo_order: &mut Vec<&'a Package>,
         visited: &mut HashSet<String>,
         cur_pkg_full_name: &String,
-        top: bool,
```
Comment (lijunchen) [inline crates/moonbuild/src/gen/gen_runtest.rs:236]: The purpose of `top` is to limit `test_import` only propagation once.

Comment (Young-Flash) [reply crates/moonbuild/src/gen/gen_runtest.rs:236]: with_testimport

File: crates/moon/tests/test_cases/blackbox_failed_test.in/C/moon.pkg.json [hunk 8]
```diff
@@ -0,0 +1,3 @@
+{
+  "import": {},
```
Comment (lijunchen) [inline crates/moon/tests/test_cases/blackbox_failed_test.in/C/moon.pkg.json]: clean up trailing comma

File: crates/moon/tests/test_cases/blackbox_success_test.in/C/moon.pkg.json [hunk 9]
```diff
@@ -0,0 +1,3 @@
+{
+  "import": {},
```
Comment (lijunchen) [inline crates/moon/tests/test_cases/blackbox_success_test.in/C/moon.pkg.json]: ditto

File: crates/moon/tests/test_cases/mod.rs [hunk 10]
```diff
@@ -4128,3 +4123,67 @@ fn test_third_party() {
...
+            [
+                "test",
+                "-p",
+                "username/hello/A",
+                "-f",
+                "hello_bbtest.mbt",
+                "-i",
+                "0",
+                "--dry-run",
+                "--nostd",
```
Comment (lijunchen) [inline crates/moon/tests/test_cases/mod.rs:4142]: add --sort-input

File: crates/moonbuild/src/gen/gen_runtest.rs [hunk 11]
```diff
@@ -362,6 +447,52 @@ pub fn gen_link_underscore_test(
...
+                .artifact
+                .with_file_name(format!("{}.blackbox_test.core", cur_pkg.last_name()))
+        } else {
+            cur_pkg.artifact.with_extension("core")
+        };
+        core_deps.push(d.display().to_string());
+    }
+
+    // add the cur pkg .core in link-core stage
+    core_deps.push(
```
Comment (lijunchen) [inline crates/moonbuild/src/gen/gen_runtest.rs]: need fix core order

**Pull request merged**